code/modules/economy/economy_machinery/eftpos.dm

```diff
@@ -0,0 +1,218 @@
+	linked_account = GLOB.station_money_database.vendor_account
+	return ..()
+
+
+/obj/item/eftpos/proc/reconnect_database()
+	account_database = GLOB.station_money_database
+
+/obj/item/eftpos/attack_self(mob/user)
+	ui_interact(user)
+
+/obj/item/eftpos/attackby(obj/O, mob/user, params)
+	if(istype(O, /obj/item/card/id))
+		//attempt to connect to a new db, and if that doesn't work then fail
+		if(!account_database)
+			reconnect_database()
+		if(account_database)
+			if(linked_account)
+				scan_card(O, user)
+				SStgui.update_uis(src)
+			else
+				to_chat(user, "[bicon(src)]<span class='warning'>Unable to connect to linked account.</span>")
+		else
+			to_chat(user, "[bicon(src)]<span class='warning'>Unable to connect to accounts database.</span>")
+	else
+		return ..()
+
+/obj/item/eftpos/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.inventory_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "EFTPOS", name, 800, 300, master_ui, state)
+		ui.open()
+
+/obj/item/eftpos/ui_data(mob/user)
+	var/list/data = list()
+	data["machine_name"] = machine_name
+	data["transaction_locked"] = transaction_locked
+	data["transaction_paid"] = transaction_paid
+	data["transaction_purpose"] = transaction_purpose
+	data["transaction_amount"] = transaction_amount
+	data["linked_account"] = linked_account ? linked_account.account_name : null
+	return data
+
+/obj/item/eftpos/ui_act(action, list/params, datum/tgui/ui)
+	if(..())
+		return
+
+	. = TRUE
+
+	var/mob/living/user = ui.user
+
+	switch(action)
+		if("change_code")
+			var/attempt_code = input("Re-enter the current EFTPOS access code", "Confirm old EFTPOS code") as num
+			if(attempt_code == access_code)
+				var/trycode = input("Enter a new access code for this device (4 digits, numbers only)", "Enter new EFTPOS code") as num
+				if(trycode < 1000 || trycode > 9999)
+					alert("That is not a valid code!")
+					return
+				access_code = trycode
+
+				print_reference()
+			else
+				to_chat(user, "[bicon(src)]<span class='warning'>Incorrect code entered.</span>")
+		if("link_account")
+			if(!account_database)
+				reconnect_database()
+			if(account_database)
+				var/attempt_account_num = input("Enter account number to pay EFTPOS charges into", "New account number") as num
+				var/attempt_pin = input("Enter pin code", "Account pin") as num
+				var/datum/money_account/target_account = GLOB.station_money_database.find_user_account(attempt_account_num, include_departments = TRUE)
+				if(!target_account)
+					for(var/department_key in GLOB.station_money_database.department_accounts)
+						var/datum/money_account/department_account = GLOB.station_money_database.department_accounts[department_key]
+						if(department_account.account_number == attempt_account_num)
+							target_account = department_account
+				if(!Adjacent(user))
+					return
```

Comment:
```suggestion
				if(!Adjacent(user))
					return
				var/datum/money_account/target_account = GLOB.station_money_database.find_user_account(attempt_account_num, include_departments = TRUE)
				if(!target_account)
					for(var/department_key in GLOB.station_money_database.department_accounts)
						var/datum/money_account/department_account = GLOB.station_money_database.department_accounts[department_key]
						if(department_account.account_number == attempt_account_num)
							target_account = department_account
```
Best to move the adjacent check right after the input.

Reply:
done